File: stack/20.valid-parentheses.go
package main

import "fmt"

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start
func isValid(s string) bool {
	pairs := map[string]string{
		")": "(",
		"}": "{",
		"]": "[",
	}
	stack := []string{}
	for _, i := range s {
		d := string(i)
		if _, ok := pairs[d]; !ok {
			stack = append(stack, d)
			continue
		}

		if len(stack) > 0 && stack[len(stack)-1] == pairs[d] {
			stack = stack[0 : len(stack)-1]
			continue
		}
		stack = append(stack, d)

	}
	if len(stack) == 0 {
		return true
	}
	return false
}

// @lc code=end

func main() {
	s := "[]"
	fmt.Println(isValid(s))
}

File: stack/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#
from typing import List


# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {")": "(", "]": "[", "}": "{", "(": ")", "[": "]", "{": "}"}
        stack: List[str] = []
        for i in s:
            if not stack:
                stack.append(i)
                continue
            if pairs[i] == stack[-1]:
                stack.pop()
            else:
                stack.append(i)
        if stack:
            return False
        return True


# @lc code=end
